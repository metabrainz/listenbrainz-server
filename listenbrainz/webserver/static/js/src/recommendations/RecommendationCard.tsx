import * as React from "react";
import {
  faAngry,
  faFrown,
  faMeh,
  faSmileBeam,
  faGrinStars,
} from "@fortawesome/free-solid-svg-icons";

import {
  faThumbsUp as faThumbsUpRegular,
  faAngry as faAngryRegular,
  faFrown as faFrownRegular,
  faSmileBeam as faSmileBeamRegular,
  faGrinStars as faGrinStarsRegular,
} from "@fortawesome/free-regular-svg-icons";

import { get as _get } from "lodash";
import { IconProp } from "@fortawesome/fontawesome-svg-core";
import { IconDefinition } from "@fortawesome/fontawesome-common-types"; // eslint-disable-line import/no-unresolved
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import RecommendationControl from "./RecommendationControl";

import { getArtistLink, getTrackLink } from "../utils";
import Card from "../components/Card";
import APIService from "../APIService";

export type RecommendationCardProps = {
  apiUrl: string;
  recommendation: Recommendation;
  playRecommendation: (recommendation: Recommendation) => void;
  currentFeedback: RecommendationFeedBack | null;
  updateFeedback: (
    recordingMbid: string,
    rating: RecommendationFeedBack | null
  ) => void;
  className?: string;
  isCurrentUser: Boolean;
  currentUser?: ListenBrainzUser;
  newAlert: (
    alertType: AlertType,
    title: string,
    message: string | JSX.Element
  ) => void;
};

export default class RecommendationCard extends React.Component<
  RecommendationCardProps
> {
  APIService: APIService;
  playRecommendation: (recommendation: Recommendation) => void;

  constructor(props: RecommendationCardProps) {
    super(props);

    this.APIService = new APIService(
      props.apiUrl || `${window.location.origin}/1`
    );

    this.playRecommendation = props.playRecommendation.bind(
      this,
      props.recommendation
    );
  }

  submitFeedback = async (rating: RecommendationFeedBack) => {
    const {
      recommendation,
      currentUser,
      isCurrentUser,
      updateFeedback,
    } = this.props;
    if (isCurrentUser && currentUser?.auth_token) {
      const recordingMBID = _get(
        recommendation,
        "track_metadata.additional_info.recording_mbid"
      );
      try {
        const status = await this.APIService.submitRecommendationFeedback(
          currentUser.auth_token,
          recordingMBID,
          rating
        );
        if (status === 200) {
          updateFeedback(recordingMBID, rating);
        }
      } catch (error) {
        this.handleError(
          `Error while submitting recommendation feedback - ${error.message}`
        );
      }
    }
  };

  deleteFeedback = async () => {
    const {
      recommendation,
      currentUser,
      isCurrentUser,
      updateFeedback,
    } = this.props;
    if (isCurrentUser && currentUser?.auth_token) {
      const recordingMBID = _get(
        recommendation,
        "track_metadata.additional_info.recording_mbid"
      );
      try {
        const status = await this.APIService.deleteRecommendationFeedback(
          currentUser.auth_token,
          recordingMBID
        );
        if (status === 200) {
          updateFeedback(recordingMBID, null);
        }
      } catch (error) {
        this.handleError(
          `Error while deleting recommendation feedback - ${error.message}`
        );
      }
    }
  };

  handleError = (error: string | Error, title?: string): void => {
    const { newAlert } = this.props;
    if (!error) {
      return;
    }
    newAlert(
      "danger",
      title || "Error",
      typeof error === "object" ? error.message : error
    );
  };

  render() {
    const {
      currentFeedback,
      recommendation,
      className,
      isCurrentUser,
      currentUser,
    } = this.props;
    let icon: IconDefinition;
    let text: string;
    switch (currentFeedback) {
      case "hate":
        icon = faAngry;
        text = "Hate";
        break;
      case "dislike":
        icon = faFrown;
        text = "Dislike";
        break;
      case "like":
        icon = faSmileBeam;
        text = "Like";
        break;
      case "love":
        icon = faGrinStars;
        text = "Love";
        break;
      default:
        icon = faThumbsUpRegular;
        text = "Like";
        break;
    }

    return (
      <Card
        onDoubleClick={this.playRecommendation}
        className={`recommendation-card row ${className}`}
      >
        <div className="track-details">
          <div title={recommendation.track_metadata.track_name}>
            {getTrackLink(recommendation)}
          </div>
          <small
            className="text-muted"
            title={recommendation.track_metadata.artist_name}
          >
            {getArtistLink(recommendation)}
          </small>
        </div>
        {isCurrentUser && currentUser?.auth_token && (
          <div className="recommendation-controls">
            <button
              className={`btn ${currentFeedback}`}
              id="recommendationControlsDropdown"
              data-toggle="dropdown"
              aria-haspopup="true"
              aria-expanded="true"
              type="button"
            >
              <FontAwesomeIcon icon={icon as IconProp} /> {text}
            </button>
            <ul
              className="dropdown-menu dropdown-menu-right"
              aria-labelledby="recommendationControlsDropdown"
            >
              <RecommendationControl
                iconHover={faAngry}
                icon={faAngryRegular}
                title="I never want to hear this again!"
                action={() =>
                  currentFeedback === "hate"
                    ? this.deleteFeedback()
                    : this.submitFeedback("hate")
                }
                cssClass={`hate ${
                  currentFeedback === "hate" ? "selected" : ""
                }`}
              />
              <RecommendationControl
                iconHover={faFrown}
                icon={faFrownRegular}
                title="I don't like this!"
                action={() =>
                  currentFeedback === "dislike"
                    ? this.deleteFeedback()
                    : this.submitFeedback("dislike")
                }
                cssClass={`dislike ${
                  currentFeedback === "dislike" ? "selected" : ""
                }`}
              />
              <RecommendationControl
                iconHover={faSmileBeam}
                icon={faSmileBeamRegular}
                title="I like this!"
                action={() =>
                  currentFeedback === "like"
                    ? this.deleteFeedback()
                    : this.submitFeedback("like")
                }
                cssClass={`like ${
                  currentFeedback === "like" ? "selected" : ""
                }`}
              />
              <RecommendationControl
                iconHover={faGrinStars}
                icon={faGrinStarsRegular}
                title="I really love this!"
                action={() =>
                  currentFeedback === "love"
                    ? this.deleteFeedback()
                    : this.submitFeedback("love")
                }
                cssClass={`love ${
                  currentFeedback === "love" ? "selected" : ""
                }`}
              />
            </ul>
          </div>
        )}
      </Card>
    );
  }
}
