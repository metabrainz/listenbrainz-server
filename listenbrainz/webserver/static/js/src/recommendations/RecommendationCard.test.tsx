import * as React from "react";
import { mount, shallow } from "enzyme";

import RecommendationCard, {
  RecommendationCardProps,
} from "./RecommendationCard";

// Font Awesome generates a random hash ID for each icon everytime.
// Mocking Math.random() fixes this
// https://github.com/FortAwesome/react-fontawesome/issues/194#issuecomment-627235075
jest.spyOn(global.Math, "random").mockImplementation(() => 0);

const recommendation: Recommendation = {
  listened_at: 0,
  track_metadata: {
    artist_name: "Kishore Kumar",
    track_name: "Ek chatur naar",
    additional_info: {
      release_mbid: "yyyy",
      artist_mbids: ["xxxx"],
    },
  },
};

const props: RecommendationCardProps = {
  recommendation,
  isCurrentUser: true,
  currentUser: { auth_token: "lalala", name: "test" },
  playRecommendation: () => {},
};

describe("RecommendationCard", () => {
  it("renders correctly for recommendations", () => {
    const wrapper = mount<RecommendationCard>(
      <RecommendationCard {...props} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
