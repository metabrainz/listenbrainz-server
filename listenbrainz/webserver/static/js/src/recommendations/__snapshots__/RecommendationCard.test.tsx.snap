// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendationCard renders correctly for recommendations 1`] = `
<RecommendationCard
  currentUser={
    Object {
      "auth_token": "lalala",
      "name": "test",
    }
  }
  isCurrentUser={true}
  playRecommendation={[Function]}
  recommendation={
    Object {
      "listened_at": 0,
      "track_metadata": Object {
        "additional_info": Object {
          "artist_mbids": Array [
            "xxxx",
          ],
          "release_mbid": "yyyy",
        },
        "artist_name": "Kishore Kumar",
        "track_name": "Ek chatur naar",
      },
    }
  }
>
  <ForwardRef
    className="recommendation-card row undefined"
    onDoubleClick={[Function]}
  >
    <div
      className="recommendation-card row undefined"
      onDoubleClick={[Function]}
      style={
        Object {
          "background": "#FFFFFF",
          "border": "1px solid #EEEEEE",
          "borderRadius": "8px",
          "boxShadow": "0 1px 1px rgba(192,192,192,0.1), 0 2px 2px rgba(192,192,192,0.15), 0 4px 4px rgba(192,192,192,0.20)",
          "boxSizing": "border-box",
          "height": "100%",
        }
      }
    >
      <div
        className="col-xs-9"
      >
        <div
          className="col-xs-11"
        >
          <div
            className="track-details"
          >
            <p
              title="Ek chatur naar"
            >
              Ek chatur naar
            </p>
            <p>
              <small
                className="text-muted"
                title="Kishore Kumar"
              >
                <a
                  href="http://musicbrainz.org/artist/xxxx"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Kishore Kumar
                </a>
              </small>
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-xs-1 text-center"
      >
        <div
          className="recommendation-controls"
        >
          <FontAwesomeIcon
            aria-expanded="true"
            aria-haspopup="true"
            border={false}
            className="dropdown-toggle"
            data-toggle="dropdown"
            fixedWidth={false}
            flip={null}
            icon={
              Object {
                "icon": Array [
                  192,
                  512,
                  Array [],
                  "f142",
                  "M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z",
                ],
                "iconName": "ellipsis-v",
                "prefix": "fas",
              }
            }
            id="recommendationControlsDropdown"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          >
            <svg
              aria-expanded="true"
              aria-haspopup="true"
              aria-hidden="true"
              className="svg-inline--fa fa-ellipsis-v fa-w-6 dropdown-toggle"
              data-icon="ellipsis-v"
              data-prefix="fas"
              data-toggle="dropdown"
              focusable="false"
              id="recommendationControlsDropdown"
              role="img"
              style={Object {}}
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </FontAwesomeIcon>
          <ul
            aria-labelledby="recommendationControlsDropdown"
            className="dropdown-menu dropdown-menu-right"
          >
            <span
              className="angry"
              title="I never want to hear this again!"
            >
              ðŸ˜ 
            </span>
            <span
              className="disappoint"
              title="I don't like this!"
            >
              ðŸ˜ž
            </span>
            <span
              className="meh"
              title="This is a bad recommendation!"
            >
              ðŸ˜‘
            </span>
            <span
              className="smile"
              title="I like this!"
            >
              ðŸ™‚
            </span>
            <span
              className="laugh"
              title="I really love this!"
            >
              ðŸ˜„
            </span>
          </ul>
        </div>
      </div>
    </div>
  </ForwardRef>
</RecommendationCard>
`;
